File: app/src/main/java/eu/qm/fiszki/AlarmReceiver.java
package eu.qm.fiszki;

import android.app.Activity;
import android.app.AlarmManager;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Build;
import androidx.core.app.NotificationCompat;

import eu.qm.fiszki.activity.CheckActivity;

public class AlarmReceiver extends BroadcastReceiver {

    private static final String CHANNEL_ID = "fiszki_notifications";
    private static final int PENDING_INTENT_FLAGS =
            PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE;

    @Override
    public void onReceive(Context context, Intent intent) {
        createNotificationChannel(context);

        long[] vibrate = new long[]{0, 100, 0, 100};
        Bitmap icon = BitmapFactory.decodeResource(context.getResources(), R.mipmap.ic_launcher);
        PendingIntent pi = PendingIntent.getActivity(context, 69, new Intent(context,
                CheckActivity.class), PENDING_INTENT_FLAGS);
        NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(context, CHANNEL_ID);
        mBuilder.setLargeIcon(icon);
        mBuilder.setSmallIcon(R.mipmap.ic_stat_f);
        mBuilder.setStyle(new NotificationCompat.BigTextStyle().bigText(context.getString(R.string.notification_message)));
        mBuilder.setContentTitle(context.getString(R.string.notification_title));
        mBuilder.setContentText(context.getString(R.string.notification_message));
        mBuilder.setContentIntent(pi);
        mBuilder.setAutoCancel(true);
        mBuilder.setVibrate(vibrate);
        NotificationManager nm =
                (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        nm.notify(100, mBuilder.build());
    }

    public void start(Activity activity) {
        Intent alarmIntent = new Intent(activity.getBaseContext(), AlarmReceiver.class);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(activity.getBaseContext(), 0, alarmIntent, PENDING_INTENT_FLAGS);
        AlarmManager manager = (AlarmManager) activity.getSystemService(Context.ALARM_SERVICE);
        long time = (1000 * 60 * timeSetter(activity));
        manager.setRepeating(AlarmManager.RTC_WAKEUP, System.currentTimeMillis() + time, time, pendingIntent);
    }

    public void close(Context context) {
        Intent alarmIntent = new Intent(context, AlarmReceiver.class);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context, 0, alarmIntent, PENDING_INTENT_FLAGS);
        AlarmManager manager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        manager.cancel(pendingIntent);
    }

    private void createNotificationChannel(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    CHANNEL_ID,
                    context.getString(R.string.notification_title),
                    NotificationManager.IMPORTANCE_DEFAULT);
            channel.setDescription(context.getString(R.string.notification_message));
            channel.enableVibration(true);
            channel.setVibrationPattern(new long[]{0, 100, 0, 100});
            NotificationManager nm =
                    (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
            nm.createNotificationChannel(channel);
        }
    }

    private int timeSetter(Activity activity){
        LocalSharedPreferences localSharedPreferences= new LocalSharedPreferences(activity);
        if (localSharedPreferences.getNotificationPosition()==1){
            return 1;
        }else if(localSharedPreferences.getNotificationPosition()==2){
            return 5;
        }else if(localSharedPreferences.getNotificationPosition()==3){
            return 15;
        }else if(localSharedPreferences.getNotificationPosition()==4){
            return 30;
        }else{
            return 60;
        }
    }
}

File: app/src/main/java/eu/qm/fiszki/AlarmReceiver.kt
package eu.qm.fiszki

import android.app.AlarmManager
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import androidx.core.app.NotificationCompat
import eu.qm.fiszki.activity.CheckActivity
import java.util.Calendar

class AlarmReceiver : BroadcastReceiver() {

    companion object {
        private const val CHANNEL_ID = "fiszki_notifications"
        private const val PENDING_INTENT_FLAGS =
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        private const val REQUEST_CODE = 42

        fun scheduleNext(context: Context) {
            val prefs = LocalSharedPreferences(context)
            if (!prefs.notificationEnabled) return

            val hour = prefs.notificationHour
            val minute = prefs.notificationMinute
            val enabledDays = prefs.notificationDays

            if (enabledDays.isEmpty()) return

            val now = Calendar.getInstance()
            val alarm = Calendar.getInstance().apply {
                set(Calendar.HOUR_OF_DAY, hour)
                set(Calendar.MINUTE, minute)
                set(Calendar.SECOND, 0)
                set(Calendar.MILLISECOND, 0)
            }

            // Find the next enabled day
            // Calendar days: Sun=1, Mon=2, ..., Sat=7
            // Our days: Mon=1, Tue=2, ..., Sun=7
            fun calendarDayToOurDay(calDay: Int): Int = if (calDay == Calendar.SUNDAY) 7 else calDay - 1

            fun ourDayToCalendarDay(ourDay: Int): Int = if (ourDay == 7) Calendar.SUNDAY else ourDay + 1

            // If today's time has passed, start searching from tomorrow
            if (alarm.before(now) || alarm == now) {
                alarm.add(Calendar.DAY_OF_YEAR, 1)
            }

            // Search up to 7 days ahead for an enabled day
            for (i in 0 until 7) {
                val ourDay = calendarDayToOurDay(alarm.get(Calendar.DAY_OF_WEEK))
                if (enabledDays.contains(ourDay.toString())) break
                alarm.add(Calendar.DAY_OF_YEAR, 1)
            }

            val intent = Intent(context, AlarmReceiver::class.java)
            val pi = PendingIntent.getBroadcast(context, REQUEST_CODE, intent, PENDING_INTENT_FLAGS)
            val am = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
            am.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, alarm.timeInMillis, pi)
        }

        fun cancel(context: Context) {
            val intent = Intent(context, AlarmReceiver::class.java)
            val pi = PendingIntent.getBroadcast(context, REQUEST_CODE, intent, PENDING_INTENT_FLAGS)
            val am = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
            am.cancel(pi)
        }
    }

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action == Intent.ACTION_BOOT_COMPLETED) {
            val prefs = LocalSharedPreferences(context)
            if (prefs.notificationEnabled) {
                scheduleNext(context)
            }
            return
        }

        // Alarm fired — show notification, then schedule next
        createNotificationChannel(context)

        val icon = BitmapFactory.decodeResource(context.resources, R.mipmap.ic_launcher)
        val pi = PendingIntent.getActivity(
            context, 69,
            Intent(context, CheckActivity::class.java), PENDING_INTENT_FLAGS
        )
        val builder = NotificationCompat.Builder(context, CHANNEL_ID).apply {
            setLargeIcon(icon)
            setSmallIcon(R.mipmap.ic_stat_f)
            setStyle(NotificationCompat.BigTextStyle().bigText(context.getString(R.string.notification_message)))
            setContentTitle(context.getString(R.string.notification_title))
            setContentText(context.getString(R.string.notification_message))
            setContentIntent(pi)
            setAutoCancel(true)
        }
        val nm = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        nm.notify(100, builder.build())

        scheduleNext(context)
    }

    private fun createNotificationChannel(context: Context) {
        val channel = NotificationChannel(
            CHANNEL_ID,
            context.getString(R.string.notification_title),
            NotificationManager.IMPORTANCE_DEFAULT
        ).apply {
            description = context.getString(R.string.notification_message)
        }
        val nm = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        nm.createNotificationChannel(channel)
    }
}
